subprojects/distributions/binary-compatibility.gradle

```diff
@@ -0,0 +1,143 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+
+import japicmp.model.JApiChangeStatus
+import me.champeau.gradle.japicmp.JapicmpTask
+import org.gradle.api.attributes.Attribute
+import org.gradle.binarycompatibility.rules.*
+import org.gradle.binarycompatibility.transforms.*
+
+buildscript {
+    repositories {
+        maven {
+            url "https://plugins.gradle.org/m2/"
+        }
+    }
+    dependencies {
+        classpath 'me.champeau.gradle:japicmp-gradle-plugin:0.2.1'
+    }
+}
+
+repositories {
+    ivy {
+        name 'Gradle distributions'
+        url 'https://services.gradle.org'
+        layout 'pattern', {
+            ivy '[module]-ivy.xml'
+            artifact '/distributions/[module]-[revision]-bin(.[ext])'
+        }
+    }
+}
+
+configurations {
+    baseline
+}
+
+def ARTIFACT_TYPE = Attribute.of('artifactType', String)
+Set publicProjects = pluginProjects + implementationPluginProjects + publicGroovyProjects - [project(':smokeTest'), project(':soak')]
+def projects = publicProjects.collect { it.archivesBaseName - 'gradle-' }
+def baselineVersion = project.findProperty('baselineVersion')?:releasedVersions.mostRecentFinalRelease
+
+dependencies {
+    baseline "gradle:gradle:${baselineVersion}@zip"
+
+    // This transform takes the Gradle zip distribution,
+    // and unzips the Gradle jar files that it contains in a directory
+    registerTransform {
+        from.attribute(ARTIFACT_TYPE, 'zip')
+        to.attribute(ARTIFACT_TYPE, 'gradle-libs-dir')
+        artifactTransform(ExplodeZipAndFindJars)
+    }
+
+    projects.each { projectName ->
+        // This transform uses the result of the exploded zip extraction
+        // and returns a single jar file based on the lookup project name
+        registerTransform {
+            from.attribute(ARTIFACT_TYPE, 'gradle-libs-dir')
+            to.attribute(ARTIFACT_TYPE, projectName)
+            artifactTransform(FindGradleJar) {
+                params(projectName)
+            }
+        }
+    }
+}
+
+def acceptedViolations = loadBreakingChanges(project, baselineVersion)
+def baselineConfigurations = []
+projects.each { projectName ->
+    def appendixName = projectName.split('-')*.capitalize().join('')
+    baselineConfigurations << configurations.create("japicmp-baseline-${appendixName}") {
+        extendsFrom configurations.baseline
+        attributes.attribute(ARTIFACT_TYPE, projectName)
+    }
+}
+
+task checkBinaryCompatibility(type: JapicmpTask) {
+    dependsOn publicProjects.jar
+    inputs.files(publicProjects.jar.archivePath) // this shouldn't be necessary, something to debug on the plugin!
+    inputs.file(project.file("src/changes/changes-since-${baselineVersion}.txt"))
+    inputs.property("baseline.version", baselineVersion)
+
+    // we're going to do an artifact based comparison, not classpath based
+    oldClasspath = files()
```

Comment:
We can do this. I wanted to limit the comparison to our API first. Types leaking is another beast, but we can tackle it too I guess...